lib/pure/views.nim

```diff
@@ -0,0 +1,148 @@
+## View is a type representing a range of elements in an array. In can be thought as a pointer plus a size.
+## View can be created to an arbitrary memory segment and can additionally keep a single ``ref`` object alive.
+##
+## This module defines views and several helper operations on them. All functions in this module (except for unsafe version of ``initView``) are designed to be memory safe (e.g. they raise exception on out-of-bounds accesses).
+import strutils
+
+const
+  needGcKeep = not (compileOption("gc", "boehm") or compileOption("gc", "none"))
+
+type
+  View*[T] = object
+    data: ptr T
+    length: int
+    when needGcKeep:
+      gcKeep: RootRef
+
+  ByteView* = View[byte]
+
+proc initView*[T; R: ref](data: ptr T, len: int, gcKeep: R): View[T] =
+  ## Creates a new view pointing, starting at ``data`` with length ``len``.
+  ##
+  ## ``gcKeep`` should be a pointer to the underlying GC objects that contains these items.
+  when needGcKeep:
+    return View[T](data: data, length: len, gcKeep: cast[RootRef](gcKeep))
+  else:
+    return View[T](data: data, length: len)
+
+proc initView*[T](data: ptr T, len: int): View[T] =
+  ## Creates a new view, starting at ``data`` with length ``len``.
+  ##
+  ## You need to ensure that data pointed by this view is not garbage collected while using the view.
+  when needGcKeep:
+    return View[T](data: data, length: len, gcKeep: nil)
+  else:
+    return View[T](data: data, length: len)
+
+proc initEmptyView*[T](typ: typedesc[T]): View[T] =
+  ## Creates a view of length zero.
+  return initView[T](nil, 0)
+
+proc initView[T](s: ref seq[T]): View[T] =
```

Comment:
Shouldn't this be ``var seq[T]``?

Reply:
That wouldn't be memory safe. (You can use `initView(ptr T, int)` for that)

Reply:
Couldn't you use ``GC_ref`` to make it safe?

Reply:
Yes. But I think then it needs to allocate object to make it possible to call GC_unref in destructor.